perf: avoid calling executeOnUIRuntimeSync if shared value has initial value

janicduplessis opened the pull request:
## Summary

`executeOnUIRuntimeSync` is pretty slow, especially on low end android devices, and always comes up at the top when profiling apps, especially on older RN versions.

The idea is to use the initial value if the shared value hasn't changed yet. When the shared value changes the first time we call into JS to set the dirty flag to true and then it will fall back to slow path calling `executeOnUIRuntimeSync`.

This significantly reduces calls to `executeOnUIRuntimeSync` since most shared values are not changed before being connected to a view.

## Test plan

Tested by patching this in an app with heavy reanimated usage and making sure animations still work properly.

Ran profiles and seen time spend in `executeOnUIRuntimeSync` goes from ~4000ms on RN 0.74 or ~400ms on RN 0.79 to about 0.

Before:

0.74

<img width="649" height="422" alt="image" src="https://github.com/user-attachments/assets/d52c6e3a-01fb-4223-bac4-9e3ee773cf48" />

0.79

<img width="585" height="518" alt="image" src="https://github.com/user-attachments/assets/450712d8-fa8a-4ef7-9941-8920f4116b79" />

After:

No longer in the sample profile.

janicduplessis commented:
Currently looking into making sure that in practice a race condition where the value is updated on UI thread, then read on JS before the dirty flag is passed to JS is impossible.

tjzel commented:
Hi @janicduplessis, thanks for your contribution. This is an optimization we considered some time ago, but unfortunately it has some problems and we decided not to go further with it at the time. It can break the users' assumptions.

Take a look at the following example, where the user synchronously updates the shared value:

```js
import React from 'react';
import { StyleSheet, View, Button } from 'react-native';
import { executeOnUIRuntimeSync } from 'react-native-worklets';
import Animated, {
  useAnimatedStyle,
  useSharedValue,
} from 'react-native-reanimated';

export default function Component() {
  const size = useSharedValue(100);
  const [, setRenderCount] = React.useState(0);

  const animatedStyle = useAnimatedStyle(() => ({
    width: size.value,
    height: size.value,
    backgroundColor: 'blue',
  }));

  return (
    <View style={styles.container}>
      <Animated.View style={animatedStyle} />
      <Button
        title="Toggle"
        onPress={() => {
          executeOnUIRuntimeSync(() => {
            size.value = 200;
          })();
          setRenderCount((count) => count + 1);
        }}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
});
```

With your changes this will result in the change not applied at all, because the React Native Runtime couldn't yet know that the Shared Value was dirty:

## Before 
https://github.com/user-attachments/assets/89fd0e90-9bec-4d7a-84a8-4d600c5788a0

## After 
https://github.com/user-attachments/assets/99fc80d5-04e6-4622-8928-9b6d89454229

While the example might seem a bit artificial, the same interaction could happen if a gesture changed a shared value at the same moment that a component is rerendered, losing the update from the gesture.

I think we could add a feature flag/another parameter to `useSharedValue` to allow this kind of behavior if the user is not afraid of flickers or lost updated but I'm not sure if we want to ship this into production for everyone.

However, with some more fast-memory features that are coming to `react-native-worklets` we might optimize initial reads by keeping the dirty flag in not runtime-bound state.

janicduplessis commented:
Hey @tjzel, thanks for the example. I thought the value from JS was only read during the first render (https://github.com/janicduplessis/react-native-reanimated/blob/eb246ef260702e48dbcf00e55088503b2151da14/packages/react-native-reanimated/src/createAnimatedComponent/PropsFilter.tsx#L91-L96), is it being read somewhere else on updates. It is def more likely to happen if it can happen on updates too.

janicduplessis commented:
As for a more reliable solution what do you think about storing the dirty flag on some shared object in c++ and have the host object return that? I think keeping that dirty flag logic might be better for now to avoid having to introduce locking on the actual value.

tjzel commented:
> Any ideas about this point I mentionned here? https://github.com/software-mansion/react-native-reanimated/pull/7913#issuecomment-3120430502

I can't tell from the top of my head, but the case on the video might be that a render overwrites an updated style with a stale style. Would need to look into it more to recall all the shenanigans with style initializations etc.

> Also any timeline on Synchronizable? Or should I look into implementing it without it and we could migrate this to use it later when it's ready.

It was pretty much close to completion but lacked some polish and testing. If you think it's a good idea to ship it asap then we could make it internally used in Shared Values in Reanimated 4.1.0 behind a feature flag.

janicduplessis commented:
@tjzel I got it working with your Sychronizable implementation. I was able to repro the problem with my previous implementation using your last example, and can confirm this works now.

If you are good with getting the Sychronizable feature out for internal usage in the current state I think this would work well.

I assume we might want to split this into 2 PRs:

- 745009779e636f3001f24f19ff358e8130ada350: Fixed conflicts and slightly simplified (supports only number, removed the extra js logic in synchronizable.ts as it wasn't working)
- 5bce8339ab6af0f3808cc85f6bd86929bea9035c: Use Synchronizable to implement the dirty flag

tjzel commented:
@janicduplessis I brought `Synchronizable` to its full initial implementation, could you verify if the benchmarks still check? I took the liberty to update your PR.

Btw, after some consideration this solution is still not error-free. Not locking the runtime might still get us a stale value because shared value updates are no longer "runtime transactions".

<img width="1705" height="1577" alt="transaction" src="https://github.com/user-attachments/assets/cfcdecdc-ee47-4714-893f-06406d145ff0" />

We'd probably need a more elaborate example with i.e. gesture handler to document the behavior difference it induces.

janicduplessis commented:
@tjzel Nice! Are we ok with that potential race condition? `value` is actually only used for the first render, so for a stale value to be read it would have to be a component mounting at the exact same time as a gesture causes an animated value to update for the first time.

That version of the Synchronizable might be a bit hard for me to test since the app is still on v3, and seems like things like `SerializableJSRef` do not exist. However I don't see anything there that would make it slower vs the version I tested.

tjzel commented:
I wish we could be ok, but unfortunately no, this could break some user setups. My idea is to enable this optimization behind a feature flag so that the user opts into this behavior consciously - although it's hard to come up with a good name for it.

janicduplessis commented:
Feature flag sounds good to me.

tjzel commented:
@janicduplessis weirdly the PR has closed after I merged the upstream branch. Can you open it again?
